docs/design/sql/18-data-links.md

```diff
@@ -0,0 +1,736 @@
+correct. It was proposed to use:
+
+```sql
+TEST CONNECTION /* … the rest equal to CREATE */
+```
+
+It was meant to be able to test a connection before it was created, because it
+assumed that we won’t be able to modify the data link later. We **will not
+implement this command**, because we will provide means to modify the data link.
+
+To test connections for already defined data link, we can use one of these:
+
+```sql
+ALTER DATA LINK <data link name> TEST CONNECTION;
+TEST CONNECTION <data link name>;
+```
+
+The `ALTER .. TEST` command is confusing, we’re not altering the data link in
+any way. It’s inspired by `ALTER JOB &lt;name> RESTART`, which we already
+support.
+
+The second form with `TEST` keyword introduces a new initial keyword, and it’s
+also a completely new command for the user to remember. Also the CONNECTION
+keyword isn’t use anywhere else. With `TEST DATA LINK` I also find it confusing.
+
+**I propose to not implement any dedicated command for testing the connection.
+Instead, to test a data link, one can use `SHOW RESOURCES` command.**
+
+## Use the data link in a mapping
+
+To create a mapping using an existing data link, use this command:
+
+
+```sql
+CREATE [OR REPLACE] [EXTERNAL] MAPPING [IF NOT EXISTS] <mapping name>
+[EXTERNAL NAME <resource name>]
+(  /* columns */ )
+DATA LINK <data link name> [OBJECT TYPE <object type>]
+[ OPTIONS … ]
+```
+
+That is, instead of `CONNECTOR TYPE &lt;connector type>` the user will refer to
+a data link. The connector can throw an error, if some option in the mapping
+conflicts with the options in the data link definition (e.g. a different target
+IP address).
+
+### Mapping without a connector
+
+For any new SQL connectors the only way to create a mapping will be only through
+a data link. We’ll keep the direct mapping for current connectors. The reason is
+to avoid _having two ways to do the same thing._ Secondly, the connection
+lifecycle is better defined with a data link, currently users might be surprised
+to learn that if you have a Kafka mapping without a data link, and execute 10
+INSERT commands, the engine connects and disconnects to Kafka 10 times.
+
+## SHOW commands
+
+
+```sql
+SHOW RESOURCES FOR <data link name>;
+
+SHOW DATA LINKS;
+
+SELECT *
+FROM information_schema.data_links;
+```
+
+The list of data links should also include links created in the config. In the
+`information_schema` there should be a flag for such data links.
+
+## GET_DDL system function
+
+This is an alternative to the proposed DESCRIBE command. The DESCRIBE command is
+used by SQL server and it doesn’t produce DDL, but it’s more like an
+information_schema query: a table with columns and their types.
+
+PostgreSQL uses a command-line tool `pg_dump`, it’s not possible to get the DDL
+through SQL. Oracle uses `dbms_metadata.get_ddl`, which we can mimic. Its
+arguments will be:
+
+
+* **namespace**: allowed values: ‘relation’, ‘datalink’. ‘relation’ is used for
```

Comment:
will it be `relation` or `table`? see https://hazelcast.atlassian.net/browse/HZ-2024

Reply:
Yeah, it was updated recently, I haven't visit it afterwards. `'relation'`.

Reply:
It will be `relation`. While some people might not be familiar with that term, the fact that it is used for mappings, views and even types might be confusing. In the future we might have 1st class tables, then it will be even more confusing.